### ISSUE TITLE
[BUG] SAML login nextUrl drops URL hash

### ISSUE THREAD

**Issue by markj-db** [opensearch-project/security-dashboards-plugin#831]
**Describe the bug**
[OpenSearch Dashboards encodes useful state in the `hash` component of the URL](https://www.elastic.co/guide/en/kibana/master/kibana-navigation.html#state-sync), for example:

```
https://<domain>/app/discover#/?_a=(columns:!(level,loggerCategory,messageText,shardName,tags.subDir,workspaceId))
```

Currently, the SAML component's `nextUrl` builder includes `url.path` but omits `url.hash`, dropping this state.  This creates a problem when sharing links to a particular view without first creating a shortlink, in that the state of that view is lost.  For the example above, the redirect would be:

```
https://<domain>/auth/saml/login?nextUrl=%2Fapp%2Fdiscover
```

and after authentication, the user would land at:

```
https://<domain>/app/discover
```

With no preselected columns.

**Expected behavior**
I believe the user expect to be redirected back to the exact URL, including hash, as they were at before the authentication redirect.  I believe it would be preferable to include `url.hash`, so the `nextUrl` would become:

```
https://<domain>/auth/saml/login?nextUrl=%2Fapp%2Fdiscover%23%2F%3F_a%3D%28columns%3A%21%28level%2CloggerCategory%2CmessageText%2CshardName%2Ctags.subDir%2CworkspaceId%29%29
```

**Additional context**
 * https://github.com/opensearch-project/security-dashboards-plugin/blob/afd64971de308b49622640cbc194d717cc802a17/server/auth/types/saml/saml_auth.ts#L57
 * https://nodejs.org/dist/latest-v8.x/docs/api/url.html#url_url_strings_and_url_objects
 * https://www.elastic.co/guide/en/kibana/master/kibana-navigation.html#state-sync

**Comment by wkruse** [comment]
This seems like a general issue. The same happens when OpenID is used. The `nextUrl` parameter drops the URL hash. So for example if you have a dashboard open and your login session expires, you are redirected to `/app/dashboards` (Dashboards overview). Or if you have a visualization open and you login session expires, you are redirected to `/app/visualize` (Visualizations overview).

Also enabling `state:storeInSessionStorage` in the Advanced Settings does **not** help, at it changes the URL hash.

Similar issue #926.

After security dashboard plugin is enabled, the usability of OpenSearch Dashboards suffers a lot, as the end users do **not** understand, why they redirected to unexpected locations after re-login.

**Comment by peternied** [comment]
Thanks for commenting on this issue, there is a pull request to save the hash from the url on redirect, https://github.com/opensearch-project/security-dashboards-plugin/pull/1001 that I believe would address this issue as well as several impacted by the same root cause.

**Comment by davidlago** [comment]
This looks like it was successfully backported/released for 2.2, closing
